src/main/kotlin/interactor/GetCityHasMoreSavingsPerMonthInteractor.kt

```diff
@@ -1,73 +1,66 @@
 package interactor
 
 import model.CityEntity
+import model.FoodPrices
+import model.RealEstatesPrices
 
 
 class GetCityHasMoreSavingsPerMonthInteractor(
     private val dataSource: CostOfLivingDataSource,
 ) {
 
-
-    fun execute(): CityEntity? {
-        return dataSource.getAllCitiesData()
-            .asSequence()
-            .filter(::excludeNullFoodPrices)
-            .filter(::excludeNullApartment)
-            .filter(::excludeNullAverageMonthly)
-            .filter(::excludeTransportationsPrices)
-            .maxByOrNull(::countSavingsPerMonth)
+    fun execute(): CityEntity {
+        return dataSource
+            .getAllCitiesData()
+            .filter(::excludeNullFoodPricesAndApartmentAndSalary)
+            .maxByOrNull(::countSavingsPerMonth)?: throw  Exception("no city with the highest savings")
     }
-
-
-}
-
-
-private fun excludeNullFoodPrices(city: CityEntity): Boolean {
-    return city.foodPrices.loafOfFreshWhiteBread500g != null
-            && city.foodPrices.localCheese1kg != null
-            && city.foodPrices.beefRound1kgOrEquivalentBackLegRedMeat != null
-            && city.foodPrices.chickenFillets1kg != null
-            && city.foodPrices.riceWhite1kg != null
-
 }
 
-private fun excludeNullApartment(city: CityEntity): Boolean {
-    return city.realEstatesPrices.apartment3BedroomsInCityCentre != null
-            && city.realEstatesPrices.apartment3BedroomsOutsideOfCentre != null
 
+private fun excludeNullFoodPricesAndApartmentAndSalary(city: CityEntity): Boolean {
+    return excludeNullFoodPrices(city.foodPrices) &&
+            excludeNullApartmentPriceAndSalary( city.realEstatesPrices,
+                city.averageMonthlyNetSalaryAfterTax )
 }
 
-private fun excludeTransportationsPrices(city: CityEntity): Boolean {
-    return city.transportationsPrices.gasolineOneLiter == null
-            && city.transportationsPrices.monthlyPassRegularPrice == null
-            && city.transportationsPrices.taxi1kmNormalTariff == null
-            && city.transportationsPrices.taxi1hourWaitingNormalTariff == null
-            && city.transportationsPrices.oneWayTicketLocalTransport == null
-            && city.transportationsPrices.taxiStartNormalTariff == null
-
+private fun excludeNullFoodPrices(foodPrices: FoodPrices): Boolean {
+    return foodPrices.let {
+                it.localCheese1kg != null &&
+                it.loafOfFreshWhiteBread500g != null &&
+                it.beefRound1kgOrEquivalentBackLegRedMeat != null &&
+                it.chickenFillets1kg != null &&
+                it.riceWhite1kg != null
```

Comment:
you're not checking eggs prices.
```suggestion
                it.riceWhite1kg != null &&
                it.eggsRegular12 != null
```

Reply:
The family does not need eggs.